File: Ex-22/c++/Item.cpp
#include "Item.hpp"

Item::Item(const string &nome, const double &valorUnitario, const double &quantidade) : nome(nome), valorUnitario(valorUnitario), quantidade(quantidade)
{
}

Item::~Item()
{
}

void Item::setNome(const string &nome)
{
  this->nome = nome;
}

const string &Item::getNome() const
{
  return this->nome;
}

void Item::setValorUnitario(const double &valorUnitario)
{
  this->valorUnitario = valorUnitario;
}

const double &Item::getValorUnitario() const
{
  return this->valorUnitario;
}

void Item::setQuantidade(const double &quantidade)
{
  this->quantidade = quantidade;
}

const double &Item::getQuantidade() const
{
  return this->quantidade;
}

const double &Item::getValorTotal() const
{
  return this->valorUnitario * this->quantidade;
}

bool Item::operator==(const Item item) const
{
  return this->getNome() == item.getNome() &&
         this->getValorUnitario() == item.getValorUnitario() &&
         this->getQuantidade() == item.getQuantidade();
}

File: Ex-22/ex_22_java/Item.java
public class Item {
    private String nome;
    private double valorUnitario;
    private int quantidade;

    /**
     * Construtor que inicializa todos os atributos do item.
     * @param nome Nome do produto.
     * @param valorUnitario Preço de uma unidade do produto.
     * @param quantidade Quantas unidades estão sendo compradas.
     */
    public Item(String nome, double valorUnitario, int quantidade) {
        this.nome = nome;
        this.valorUnitario = valorUnitario;
        this.quantidade = quantidade;
    }

    /**
     * Retorna o nome do item.
     */
    public String getNome() {
        return this.nome;
    }

    /**
     * Calcula e retorna o valor total para este item (preço * quantidade).
     */
    public double getValorTotal() {
        return this.valorUnitario * this.quantidade;
    }

    /**
     * (Opcional, mas útil) Sobrescreve o método toString para uma exibição amigável.
     */
    @Override
    public String toString() {
        // Usando String.format para formatar a saída de valores monetários
        return String.format("%s (Qtd: %d, Preço Unit.: R$ %.2f, Subtotal: R$ %.2f)",
                this.nome, this.quantidade, this.valorUnitario, this.getValorTotal());
    }
}